### PR TITLE
MM-28017 badge on the pinned messages header icon.

### LINKED ISSUE #16226: Badge on the pinned messages header icon
Summary: Add a badge to the pinned post icon to inform users of new pinned posts they haven't read

Note this is blocked on this PR: https://github.com/mattermost/mattermost-webapp/pull/7030

#### Proposed Solution

  * Depends on [https://mattermost.atlassian.net/browse/MM-28017](https://mattermost.atlassian.net/browse/MM-28017|smart-link)
  * Adds a dot badge next to the pinned icon
  * Uses the ‘Active’ theme color for the badge 
  * If there's new pinned posts in the channel since the last time you opened the pinned post view (RHS), show the dot. Note that when a server upgrades to the version with this change, only show the dot for new pinned posts after the upgrade (not existing pinned posts before the server upgrade). This is to avoid the dot appearing in every channel once a server is upgraded.

[Figma design file](https://www.figma.com/file/7385m83jhgalrcnSRl3H43/MM-24067-Pinned-Messages-Badge?node-id=0%3A1)

[Figma clickable prototype](https://www.figma.com/proto/7385m83jhgalrcnSRl3H43/MM-24067-Pinned-Messages-Badge?node-id=23%3A54&viewport=484%2C542%2C0.5084013938903809&scaling=min-zoom)

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-24067

### PR DESCRIPTION
**Pull request by sk409**
#### Summary

Show the badge on the pinned messages header icon when a new pinned post exists.

Added API(/api/v4/channels/{channelId}/members/{memberId}/last_viewed_pinned_post_at) is executed when a user view pinned posts to update lastviewedpinnedpostat column on channelmembers table.

The flag to show the badge is retrieved via API(/api/v4/channels/{channelId}/stats).
When the flag is true then the badge is shown.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-24067
Fixes https://github.com/mattermost/mattermost-server/issues/16226

#### Related Pull Requests

- https://github.com/mattermost/mattermost-server/pull/21996

#### Screenshots

スクリーンショット 2023-01-05 22 23 35

#### Release Note

```release-note
The badge indicates new pinned posts is shown when the new one exists
```

### CODE REVIEW

File: packages/mattermost-redux/src/actions/search.ts
Comment (michelengelen) [inline packages/mattermost-redux/src/actions/search.ts:19]: @koox00 or @AshishDhama was this the API call that breaks the server?

**Review by michelengelen — COMMENTED** [review]
looking good so far ... will wait for the other devs and check again afterwards

File: packages/mattermost-redux/src/actions/search.ts
Comment (AshishDhama) [reply packages/mattermost-redux/src/actions/search.ts:19]: Yup, this is the one that breaks the server.

Comment (michelengelen) [reply packages/mattermost-redux/src/actions/search.ts:19]: OK, in this case we should wait for the fix on that to be merged and continue afterwards. @AshishDhama could you link/mention the other PR here so we can keep track? Thanks!

Comment (AshishDhama) [reply packages/mattermost-redux/src/actions/search.ts:19]: @michelengelen @ashishbhate will work on the backend but yeah I will post link as soon as we have it

Comment (hmhealey) [reply packages/mattermost-redux/src/actions/search.ts:19]: What's the ticket for that one?

Comment (AshishDhama) [reply packages/mattermost-redux/src/actions/search.ts:19]: https://mattermost.atlassian.net/browse/MM-49486

Comment (AshishDhama) [reply packages/mattermost-redux/src/actions/search.ts:19]: PR is up for that https://github.com/mattermost/mattermost-server/pull/22096

**Comment by hmhealey** [comment]
Removing the stale label from this since it's blocked on us currently

**Comment by hmhealey** [comment]
@AshishDhama Now that that PR is in, what's the followup to do here? Do we just need to change that API call to use the version that excludes the file count?

**Comment by AshishDhama** [comment]
@hmhealey Yeah, and I think we need to be careful with changes done for this.

**Comment by sk409** [comment]
I resolved conflicts.

File: packages/mattermost-redux/src/actions/search.ts
Comment (hmhealey) [inline packages/mattermost-redux/src/actions/search.ts:252]: Do we actually need this call here? What if we were to have the reducer for the channel stats watch for the existing `RECEIVED_SEARCH_PINNED_POSTS` action and then recognize that it should set `has_new_pinned_posts` to false?

I know there's the above thread on needing to use the safer version of that API call here, but we might be able to remove this followup API call entirely

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: packages/mattermost-redux/src/actions/search.ts
Comment (sk409) [reply packages/mattermost-redux/src/actions/search.ts:252]: @hmhealey
I think your idea can also achieve the goal.
I will reimplement this logic as you mentioned.

Comment (sk409) [reply packages/mattermost-redux/src/actions/search.ts]: @hmhealey 
I changed this to using RECEIVED_SEARCH_PINNED_POSTS to hide the badge.

File: packages/mattermost-redux/src/reducers/entities/channels.ts
Comment (hmhealey) [inline packages/mattermost-redux/src/reducers/entities/channels.ts:711]: Slight nitpicking, but would you mind moving this `case` above the `LOGOUT_SUCCESS` one? We generally tend to end reducers with the logout case followed by the default one

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for changing that. It's also nice to avoid the followup server request there becaise there's some weird race conditions we can run into by doing that in a multi-server setup.

File: packages/mattermost-redux/src/actions/search.ts
Comment (hmhealey) [reply packages/mattermost-redux/src/actions/search.ts]: @AshishDhama Resolving this since we're no longer making that extra request at all

File: packages/mattermost-redux/src/reducers/entities/channels.ts
Comment (sk409) [reply packages/mattermost-redux/src/reducers/entities/channels.ts:709]: @hmhealey 
I fixed this as you mentioned.

**Review by hmhealey — APPROVED** [review]
Thanks for changing that! This looks good to me now

**Comment by hmhealey** [comment]
@esethna I assigned this to you for PM Review since you'd created the original ticket, but let me know if I should re-assign to someone else

**Review by esethna — APPROVED** [review]
Thanks, will defer to QA. @furqanmlk please ensure we test these cases in the ticket description:

- If there's new pinned posts in the channel since the last time you opened the pinned post view (RHS), show the dot. Note that when a server upgrades to the version with this change, only show the dot for new pinned posts after the upgrade (not existing pinned posts before the server upgrade). This is to avoid the dot appearing in every channel once a server is upgraded.

- When a user joins a channel, if there are pinned posts, show the dot until the pinned posts are viewed

**Comment by furqanmlk** [comment]
Hi @sk409, Can you please re-assign to me when the Dev review is complete in both [Webapp](https://github.com/mattermost/mattermost-webapp/pull/11907) and [Server](https://github.com/mattermost/mattermost-server/pull/21996)

**Review by AshishDhama — APPROVED** [review]
LGTM!

**Comment by AshishDhama** [comment]
@esethna @matthewbirtch Do we want this Behaviour
 1. Dot appearing even if user removes post from pinned post
 2. Even the person adding pinned post getting shown the dot.
 
 cc// @sk409 

https://user-images.githubusercontent.com/16203333/222781329-3f2e9cfd-b74d-4184-a613-a4cd71e5201e.mov

**Comment by matthewbirtch** [comment]
> @esethna @matthewbirtch Do we want this Behaviour
> 
> 1. Dot appearing even if user removes post from pinned post
> 2. Even the person adding pinned post getting shown the dot.
> 
> cc// @sk409

@AshishDhama no to both, we don't want a badge to show if something is removed and we don't want to show a badge for the user who has added the pinned post.

**Comment by AshishDhama** [comment]
@sk409 Cloud you, please do the required changes in the backend.

**Comment by hmhealey** [comment]
@furqanmlk With the monorepo merge coming today, we're going to have to wait on that actually.

@sk409 If you want to start moving these changes into the monorepo, you can rebase your server PR onto https://github.com/mattermost/mattermost-server/pull/22553 and start moving these changes into that PR. We're just wrapping up some issues with the unit tests, but otherwise, that'll be merged into master soon. The migration should hopefully go smoothly, but if you run into any issues, we have some updates to the developer docs [here](https://github.com/mattermost/mattermost-developer-documentation/pull/1223) that are going to be submitted when the monorepo merge happens

**Comment by hmhealey** [comment]
The web app now lives in the `webapp/channels` over in https://github.com/mattermost/mattermost-server. We decided to move it to be a monorepo to make both development and our CI pipelines easier, but it's unfortunately disruptive right now. We've updated [our setup docs](https://developers.mattermost.com/contribute/developer-setup/) and added [some steps for migrating to working with the monorepo](https://developers.mattermost.com/contribute/monorepo-migration-notes/).

Copying the changes over there hopefully shouldn't be too hard. There's a few changes to file paths (such as the code living in a dedicated `src` subdirectory), but most things should look the same as they do now. 

Let us know if you could use any help with this. Hopefully the process is smooth.

**Comment by M-ZubairAhmed** [comment]
Hello there! Could you please open the PR in the mattermost-server as part of the monorepo? I will be closing this PR for now. Thank you very much for your understanding and cooperation in this matter.

**Comment by sk409** [comment]
@M-ZubairAhmed 
I will reopen this PR in the mattermost-server.

**Comment by sk409** [comment]
I was busy yesterday, but today I am ready to start migrating to a monorepo.

**Comment by hmhealey** [comment]
Sounds good to me. Let us know if you need any help with that, and thanks for your continued work on this

**Pull request closed without merging**
